### PR TITLE
[FLINK-18133][docs][avro] Add documentation for the new Avro format

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

This add documentation for Avro format and add a basic framework for format pages. 

## Brief change log

- hotfix the Avro format doesn't support TINYINT and SMALLINT types.
- added format overview page
- added avro format page

## Verifying this change

- added tests for Avro format for TINYINT and SMALLINT types.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: docs/dev/table/connectors/formats/avro.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      <td>record</td>
+      <td></td>
+    </tr>
+    </tbody>
+</table>
+
+In addition to the types listed above, Flink supports reading/writing nullable types. Flink maps nullable types to Avro `union(something, null)`, where `something` is any converted Avro type from Flink type.
+
+You can refers Avro Specification for more information about Avro types: [https://avro.apache.org/docs/current/spec.html](https://avro.apache.org/docs/current/spec.html).
+
```
Comment (danny0405) [inline docs/dev/table/connectors/formats/avro.md:203]: `can refers` -> `can refer to`

File: docs/dev/table/connectors/formats/avro.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    <tr>
+      <td>ROW</td>
+      <td>record</td>
+      <td></td>
+    </tr>
+    </tbody>
+</table>
+
+In addition to the types listed above, Flink supports reading/writing nullable types. Flink maps nullable types to Avro `union(something, null)`, where `something` is any converted Avro type from Flink type.
+
```
Comment (danny0405) [inline docs/dev/table/connectors/formats/avro.md:201]: `where `something` is` -> `where `something` can be`

File: docs/dev/table/connectors/formats/avro.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+      <td>Specify what format to use, here should be 'avro'.</td>
+    </tr>
+    </tbody>
+</table>
+
+Data Type Mapping
+----------------
+
+Currently, the Avro schema is always derived from table schema, explicitly define an Avro schema is not supported yet. So here only lists the conversion from Flink type to Avro type.
+
```
Comment (danny0405) [inline docs/dev/table/connectors/formats/avro.md:96]: `explicitly define` -> `explicitly definition of `

Comment (danny0405) [reply docs/dev/table/connectors/formats/avro.md:96]: `So here only lists the conversion from Flink type to A` -> The following table shows the type mapping from ...

File: docs/dev/table/connectors/formats/avro.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+* This will be replaced by the TOC
+{:toc}
+
+The [Apache Avro](https://avro.apache.org/) format allows to read and write Avro data based on an Avro schema. Currently, the Avro schema is derived from table schema.
+
+Dependencies
+------------
+
+In order to setup the Avro format, the following table provide dependency information for both projects using a build automation tool (such as Maven or SBT) and SQL Client with SQL JAR bundles.
+
```
Comment (danny0405) [inline docs/dev/table/connectors/formats/avro.md:38]: `provide` -> `provides`

File: docs/dev/table/connectors/formats/index.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+Flink provides a set of table formats that can be used with table connectors. A table format is a storage format defines how map binary data onto table columns.
+
```
Comment (danny0405) [inline docs/dev/table/connectors/formats/index.md:28]: `how` -> `how to`

File: docs/dev/table/connectors/formats/avro.md (hunk 2 above)
Comment (wuchong) [reply docs/dev/table/connectors/formats/avro.md:201]: I changed this to 

```
where `something` is the Avro type converted from Flink type.
```

File: docs/dev/table/connectors/formats/avro.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+The [Apache Avro](https://avro.apache.org/) format allows to read and write Avro data based on an Avro schema. Currently, the Avro schema is derived from table schema.
+
+Dependencies
+------------
+
+In order to setup the Avro format, the following table provides dependency information for both projects using a build automation tool (such as Maven or SBT) and SQL Client with SQL JAR bundles.
+
+| Maven dependency   | SQL Client JAR         |
+| :----------------- | :----------------------|
+| `flink-avro`       | [Pre-bundled Hadoop](https://flink.apache.org/downloads.html#additional-components) |
```
Comment (JingsongLi) [inline docs/dev/table/connectors/formats/avro.md:41]: I think it is better to link https://ci.apache.org/projects/flink/flink-docs-master/ops/deployment/hadoop.html

Comment (wuchong) [reply docs/dev/table/connectors/formats/avro.md:41]: You are right. I have also noticed that when reviewing Kafka documentation. But I don't want to block the following format documentation. So I will create other pull request for this.

File: docs/dev/table/connectors/formats/avro.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+<div class="codetabs" markdown="1">
+<div data-lang="SQL" markdown="1">
+{% highlight sql %}
+CREATE TABLE user_behavior (
+  user_id BIGINT,
+  item_id BIGINT,
+  category_id BIGINT,
+  behavior STRING,
+  ts TIMESTAMP(3),
+) WITH (
```
Comment (danny0405) [inline docs/dev/table/connectors/formats/avro.md:57]: Oops, i just noticed that there is a redundant comma.

Comment (wuchong) [reply docs/dev/table/connectors/formats/avro.md:57]: Good catch!

**Pull request closed without merging**